src/blockrelay/compactblock.cpp

```diff
@@ -0,0 +1,1216 @@
+    : nonce(GetRand(std::numeric_limits<uint64_t>::max())), header(block)
+{
+    FillShortTxIDSelector();
+
+    if (block.vtx.empty())
+        throw std::invalid_argument(__func__ + std::string(" expects coinbase tx"));
+
+    //< Index of a prefilled tx is its diff from last index.
+    size_t prevIndex = 0;
+    prefilledtxn.push_back(PrefilledTransaction{0, *block.vtx[0]});
+    for (size_t i = 1; i < block.vtx.size(); i++)
+    {
+        const CTransaction &tx = *block.vtx[i];
+        if (inventoryKnown && !inventoryKnown->contains(tx.GetHash()))
+        {
+            prefilledtxn.push_back(PrefilledTransaction{static_cast<uint16_t>(i - (prevIndex + 1)), tx});
+            prevIndex = i;
+        }
+        else
+        {
+            shorttxids.push_back(GetShortID(tx.GetHash()));
+        }
+    }
+}
+
+void CompactBlock::FillShortTxIDSelector() const
+{
+    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
+    stream << header << nonce;
+    CSHA256 hasher;
+    hasher.Write((unsigned char *)&(*stream.begin()), stream.end() - stream.begin());
+    uint256 shorttxidhash;
+    hasher.Finalize(shorttxidhash.begin());
+    shorttxidk0 = shorttxidhash.GetUint64(0);
+    shorttxidk1 = shorttxidhash.GetUint64(1);
+}
+
+uint64_t CompactBlock::GetShortID(const uint256 &txhash) const
+{
+    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
+}
+
+void validateCompactBlock(const CompactBlock &cmpctblock)
+{
+    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
+        throw std::invalid_argument("empty data in compact block");
+
+    int32_t lastprefilledindex = -1;
+    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
+    {
+        if (cmpctblock.prefilledtxn[i].tx.IsNull())
+            throw std::invalid_argument("null tx in compact block");
+
+        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
+        if (lastprefilledindex > std::numeric_limits<uint16_t>::max())
+            throw std::invalid_argument("tx index overflows");
+
+        if (static_cast<uint32_t>(lastprefilledindex) > cmpctblock.shorttxids.size() + i)
+        {
+            // If we are inserting a tx at an index greater than our full list of shorttxids
+            // plus the number of prefilled txn we've inserted, then we have txn for which we
+            // have neither a prefilled txn or a shorttxid!
+            throw std::invalid_argument("invalid index for tx");
+        }
+    }
+}
+
+/**
+ * Handle an incoming compactblock.  The block is fully validated, and if any transactions are missing, we fall
+ * back to requesting a full block.
+ */
+
+
+bool CompactBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+{
+    if (!pfrom->CompactBlockCapable())
+    {
+        dosMan.Misbehaving(pfrom, 100);
+        return error("Compact block message received from a non compactblock node, peer=%s", pfrom->GetLogName());
+    }
```

Comment:
Is this check needed?  An attack would simply mark itself of compact block capable.  If there is no strong reason for this check, we should be tolerant in what we receive.